src/main/scala/com/rbmhtechnology/calliope/scaladsl/TransactionalEventProducer.scala

```diff
@@ -203,68 +253,140 @@ private class GraphRunnerSupervisor(graph: ProducerGraph, onFailure: FailureHand
+      copy(killSwitch = Some(killSwitch))
+
+    def withNotificationFor(target: ActorRef): Data =
+      copy(notifyOnStop = notifyOnStop :+ target)
+
+    def withoutNotifications(): Data =
+      copy(notifyOnStop = Vector.empty)
+  }
+
   def props(graph: ProducerGraph): Props =
     Props(new ProducerGraphRunner(graph))
 }
 
-private class ProducerGraphRunner(graph: ProducerGraph) extends Actor with ActorLogging with Stash {
+private class ProducerGraphRunner(graph: ProducerGraph) extends Actor
+  with ActorLogging with Stash with FSM[ProducerGraphRunner.State, ProducerGraphRunner.Data] {
 
   import ProducerGraphRunner._
   import context.dispatcher
 
   implicit val materializer = ActorMaterializer()
 
-  private def runGraph(): SourceQueueWithComplete[Unit] = {
-    val (queue, ft) = graph.run()
+  private def runGraph(): (SourceQueueWithComplete[Unit], UniqueKillSwitch) = {
+    val ((queue, ks), ft) = graph.run()
 
-    ft onComplete { result =>
-      val cause = result match {
-        case Failure(err) => err
-        case Success(_) => new UnexpectedStreamCompletionException("Transactional producer stream stopped unexpectedly")
-      }
-      self ! RestartGraph(cause)
+    ft recover {
+      case _: GraphAbortedException => Done
+    } onComplete {
+      case Failure(err) => self ! RestartGraph(err)
+      case Success(_) => self ! GraphStopped
     }
-    queue
+    (queue, ks)
   }
 
-  private var commitQueue: SourceQueueWithComplete[Unit] = _
+  startWith(Stopped, Data())
+
+  when(Stopped) {
+    case Event(RunGraph, data) =>
+      log.info("Starting transactional producer stream.")
+      val (queue, ks) = runGraph()
+      goto(Running) using data.withQueue(queue).withKillSwitch(ks) replying GraphStarted
 
-  override def preStart(): Unit = {
-    self ! StartGraph
+    case Event(StopGraph, _) =>
+      stay replying GraphStopped
   }
 
-  override def postStop(): Unit = {
-    materializer.shutdown()
-    super.postStop()
+  when(Running) {
+    case Event(RunGraph, _) =>
+      stay replying GraphStarted
+
+    case Event(StopGraph, data) =>
+      log.info("Shutting down transactional producer stream.")
+      data.killSwitch.foreach(_.abort(GraphAbortedException()))
+      goto(Stopping) using data.withNotificationFor(sender())
+
+    case Event(RestartGraph(failure), data) =>
+      log.warning("Transactional producer stream was shutdown with cause: {}. Producer will be restarted", failure)
+      val (queue, ks) = runGraph()
+      stay using data.withQueue(queue).withKillSwitch(ks)
+
+    case Event(NotifyCommit, data) =>
+      data.commitQueue.foreach(_.offer(Unit))
+      stay
   }
 
-  private def initializing: Receive = {
-    case StartGraph =>
-      log.info("Starting transactional producer.")
-      commitQueue = runGraph()
+  when(Stopping) {
+    case Event(RunGraph, _) =>
+      stash()
+      stay
+
+    case Event(StopGraph, _) =>
+      stay replying GraphStopped
```

Comment:
This gets an immediate reply even though the graph is not yet stopped. While the first stop that actually initiated the stop is still waiting for a reply.